File: scripts/t-wrapper-rust/tests/translation_wrapper_test.rs
/**
 * translation-wrapper 테스트
 * TranslationWrapper 클래스 테스트
 */

use t_wrapper_rust::{TranslationWrapper, ScriptConfig};
use anyhow::Result;
use tempfile::{tempdir, NamedTempFile};
use std::fs;
use std::path::PathBuf;

#[test]
fn process_files_한국어가_포함된_파일을_처리해야_함() -> Result<()> {
    let dir = tempdir()?;
    let file_path = dir.path().join("test.tsx");
    fs::write(&file_path, r#"function Component() {
  return <div>안녕하세요</div>;
}"#)?;

    let wrapper = TranslationWrapper::new(Some(ScriptConfig {
        source_pattern: dir.path().join("**/*.tsx").to_string_lossy().to_string(),
        dry_run: true,
        ..Default::default()
    }));

    let result = wrapper.process_files()?;
    assert!(!result.is_empty());
    Ok(())
}

#[test]
fn process_files_client_모드에서는_use_client와_useTranslation_훅을_보장해야_함() -> Result<()> {
    let dir = tempdir()?;
    let file_path = dir.path().join("client.tsx");
    fs::write(&file_path, r#"function ClientComp() {
  return <div>안녕하세요</div>;
}"#)?;

    let wrapper = TranslationWrapper::new(Some(ScriptConfig {
        source_pattern: dir.path().join("**/*.tsx").to_string_lossy().to_string(),
        dry_run: false,
        mode: Some("client".to_string()),
        ..Default::default()
    }));

    wrapper.process_files()?;
    let content = fs::read_to_string(&file_path)?;
    // TODO: 실제 AST 변환 및 코드 생성 후 검증
    // assert!(content.contains("'use client'"));
    // assert!(content.contains("useTranslation"));
    // assert!(content.contains("t("));
    Ok(())
}

#[test]
fn process_files_server_모드에서는_지정한_serverTranslationFunction으로_t_바인딩을_생성해야_함() -> Result<()> {
    let dir = tempdir()?;
    let file_path = dir.path().join("server.tsx");
    fs::write(&file_path, r#"function ServerComp() {
  return <div>안녕하세요</div>;
}"#)?;

    let wrapper = TranslationWrapper::new(Some(ScriptConfig {
        source_pattern: dir.path().join("**/*.tsx").to_string_lossy().to_string(),
        dry_run: false,
        mode: Some("server".to_string()),
        server_translation_function: Some("getServerT".to_string()),
        ..Default::default()
    }));

    wrapper.process_files()?;
    let content = fs::read_to_string(&file_path)?;
    // TODO: 실제 AST 변환 및 코드 생성 후 검증
    // assert!(content.contains("await getServerT"));
    // assert!(content.contains("const { t } ="));
    // assert!(content.contains("t("));
    Ok(())
}

File: scripts/t-wrapper/__tests__/translation-wrapper.test.ts
/**
 * translation-wrapper 테스트
 */

import { wrapTranslations } from "../wrapper";
import * as path from "path";
import { writeFile, readFile, createTempDir, removeDir } from "../utils/fs-utils";

describe("translation-wrapper", () => {
  let tempDir: string;

  beforeEach(() => {
    tempDir = createTempDir("i18n-test-");
  });

  afterEach(() => {
    removeDir(tempDir);
  });

  describe("wrapTranslations", () => {
    it("한국어가 포함된 파일을 처리해야 함", async () => {
      const testFile = path.join(tempDir, "test.tsx");
      writeFile(
        testFile,
        `function Component() {
  return <div>안녕하세요</div>;
}`
      );

      const result = await wrapTranslations({
        sourcePattern: path.join(tempDir, "**/*.tsx"),
      });
      expect(result.processedFiles.length).toBeGreaterThan(0);
    });

    it("Next.js 환경에서 client 모드일 때만 'use client'를 추가해야 함", async () => {
      const testFile = path.join(tempDir, "client.tsx");
      writeFile(
        testFile,
        `function ClientComp() {
  return <div>안녕하세요</div>;
}`
      );

      await wrapTranslations({
        sourcePattern: path.join(tempDir, "**/*.tsx"),
        mode: "client",
        framework: "nextjs",
      } as any);

      const content = readFile(testFile);
      expect(content).toMatch(/["']use client["']/);
      expect(content).toContain("useTranslation");
      expect(content).toContain("t(");
    });

    it("React 환경에서 client 모드일 때는 'use client'를 추가하지 않아야 함", async () => {
      const testFile = path.join(tempDir, "client-react.tsx");
      writeFile(
        testFile,
        `function ClientComp() {
  return <div>안녕하세요</div>;
}`
      );

      await wrapTranslations({
        sourcePattern: path.join(tempDir, "**/*.tsx"),
        mode: "client",
        framework: "react",
      } as any);

      const content = readFile(testFile);
      expect(content).not.toMatch(/["']use client["']/);
      expect(content).toContain("useTranslation");
      expect(content).toContain("t(");
    });

    it("server 모드에서는 지정한 serverTranslationFunction으로 t 바인딩을 생성해야 함", async () => {
      const testFile = path.join(tempDir, "server.tsx");
      writeFile(
        testFile,
        `function ServerComp() {
  return <div>안녕하세요</div>;
}`
      );

      await wrapTranslations({
        sourcePattern: path.join(tempDir, "**/*.tsx"),
        mode: "server",
        serverTranslationFunction: "getServerT",
      } as any);

      const content = readFile(testFile);
      expect(content).toContain("await getServerT");
      expect(content).toContain("const { t } =");
      expect(content).toContain("t(");
    });
  });
});
